Bxmsdoc 2288 Rules for Employee Rostering

gsheldon opened the pull request:
Requires SME and Peer review.

Commit:
BXMSDOC-2114: Draft main.adoc

Commit:
BXMSDOC-2114: Draft main.adoc

Commit:
BXMSDOC-2114: Draft main.adoc

Commit:
BXMSDOC-2114: adding files so far, basic stuff

Commit:
BXMSDOC-2114: adding files so far, basic stuff

Commit:
BXMSDOC-2114: adding files so far, basic stuff

Commit:
BXMSDOC-2276/2114: Completed the data object creation

Commit:
BXMSDOC-2276/2114: Completed the data object creation

Commit:
BXMSDOC-2276: added EmployeeRoster planning solution

Commit:
BXMSDOC-2276: added EmployeeRoster planning solution

Commit:
BXMSDOC-2276: added EmployeeRoster planning solution

Commit:
BXMSDOC-2276: correcting build issues

Commit:
BXMSDOC-2276: correcting build issues

Commit:
BXMSDOC-2276: correcting build issues

Commit:
BXMSDOC-2276: formatting

Commit:
BXMSDOC-2276: formatting

Commit:
BXMSDOC-2276: formatting

Commit:
BXMSDOC-2288: Added BalanceEmployeesShiftNumber procedure

Commit:
BXMSDOC-2288: Added BalanceEmployeesShiftNumber procedure

Commit:
BXMSDOC-2288: Finished adding Guided Rules

BXMSDOC-2288: uploaded screenshots

Commit:
BXMSDOC-2288: Finished adding Guided Rules

BXMSDOC-2288: uploaded screenshots

Commit:
Revert "BXMSDOC-2246-master Update REBUILT variable for Mar 2 2018 patch publish."

This reverts commit d528e7e70fcd0352cc2e14645deea2c3ff176482.

yurloc commented on docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc:
```diff
@@ -0,0 +1,32 @@
+[id='wb-employee-rostering-balance-employees-shift-rule-proc']
+= Create the `BalanceEmployeesShiftNumber` guided rule 
```
![running the employee rostering sample project in decision central - mozilla firefox_042](https://user-images.githubusercontent.com/673386/37278711-e28144aa-25e8-11e8-9263-9c9a20d9e78b.png)

This is hard to read. I'd prefer to remove the code-style formatting from within headings and also break the rule name down into individual words. This applies to all headings containing a rule name.

sterobin replied:
Agree, but not much you can do about it, given the CSS for H2's.

yurloc commented on docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc:
```diff
@@ -0,0 +1,32 @@
+[id='wb-employee-rostering-balance-employees-shift-rule-proc']
+= Create the `BalanceEmployeesShiftNumber` guided rule 
+
+The `BalanceEmployeesShiftNumber` guided rule creates a soft constraint that ensures all employees are assigned a shift, or that all employees are assigned an equal number of shifts. In the employee rostering example, this constraint is created using the guided rule editor. 
```
I think we should rephrase this to help the user better understand what the rule does. The issues I can see in the description

1. ensures all employees are assigned a shift - yes, but that's a side effect, not the main point of the rule
2. all employees are assigned an equal number of shifts - ideally yes, but that's not what the rule ensures

### How the rule works (WHEN section, condition)
For each employee it counts all his/her shift assignments. The number of shifts assigned to the employee is stored in the `$shiftCount` variable.
### How the rule works (THEN section, consequence)
Soft score is modified (penalized) with the negative value of the **square** of the shift count.

This is how the *rule* works and I'm not saying this needs to be explained. It's optional in my opinion. What's more important is to explain the *idea* of the constraint implemented by the rule. The idea is to penalize extreme shift counts by using the square as a score penalty. Example:

* 3 employees with 4, 5 and 6 shifts gives a score of -16 + -25 + -36 = -77 which is a worse total score than
* 3 employees with 5, 5 and 5 shifts, giving a total score of -25 + -25 + -25 = -75. This is a higher and therefore better soft score.

This example illustrates how the score formula, implemented by the rule, gives the solver an incentive to distribute shift assignments in a balanced way. If this were the only constraint, all employees would end up being assigned to equal number of shifts. Since, in a real world data set, this probably not possible due to other constraint, I would avoid stating that the rule "makes sure all employees are assigned an equal number of shifts". "The rule ensures that shifts are assigned in a more fair, balanced way" sounds more accurate.

@gsheldon the exact wording is up to you. I'm trying some extra details to help you understand how it works.

sterobin replied:
Search all instances of "editor" and change to "designer" when referring to "guided rules designer", "DRL designer", etc. Also things like "In the editor" should be "In the designer" etc. Per DM/BA terminology change. The only time should stay "editor" is when not referring to the designer (naturally), such as the **Expression editor** within the guided rules designer, the **Editor** tab (which will be changed in DM 7.1 anyway).

Here's the terminology change overview for DM/BA, if it's helpful: https://docs.google.com/spreadsheets/d/1kmK92hY3JJp7gZTvLzd0CawaZzvI2MYYv2uaLaQcMJc/edit#gid=1999150540

sterobin commented on docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc:
```diff
@@ -0,0 +1,32 @@
+. Add a *WHEN* condition by clicking the image:5686.png[] in the *WHEN* field.
+. Select `Employee` in the *Add a condition to the rule...* window. Click *+Ok*. 
+. Click on the `Employee` condition to modify the constraints and add the variable name `$employee`.
+. Add the *WHEN* condition `From Accumulate...`. 
```
Per IBM SG, don't retain the trailing ellipsis when referring to UI text that has it at the end. So should be "the **Add a condition to the rule** window", "condition `From Accumulate`", etc. The only exception I've found is when the ellipsis is not the final element, such as "click **(show options...)**", at the bottom of the designer. I'd search all ellipses and see if they are per the UI and can be removed. Super minor, but just sayin'. Up to you.

sterobin commented on docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc:
```diff
@@ -0,0 +1,32 @@
+. Add the *WHEN* condition `From Accumulate...`. 
+.. Above the `From Accumulate...` condition, click *click to add pattern...* and select `Number` as the fact type from the drop down list.
+.. Add the variable name `$shiftCount` to the `Number` condition.
+.. Below the `From Accumulate...` condition, click *click to add pattern...* and select the `ShiftAssignment` fact type from the drop down list.
```
drop-down list/button/etc. Ditto everywhere else this appears.

sterobin commented on docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc:
```diff
@@ -0,0 +1,32 @@
+. Add the *THEN* condition by clicking the image:5686.png[] in the *THEN* field.
+. Select `Modify Soft Score` in the *Add a new action...* window. Click *+Ok*.
+.. Type the following expression into the box: `-($shiftCount.intValue()*$shiftCount.intValue())`
+. Click *Validate* in the upper right corner to check all rule conditions are valid. If the rule validation fails, address any problems described in the error message, review all components in the rule, and try again to validate the rule until the rule passes.
```
upper-right corner/toolbar/etc. Ditto everywhere else this appears.

sterobin commented on docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc:
```diff
@@ -0,0 +1,32 @@
+.. Select `equal to` from the drop down list next to the `employee` constraint.
+.. Click the image:6191.png[edit] icon next to the drop down button to add a variable, and click *Bound variable* in the *Field value* window.
+.. Select `$employee` from the drop down list.
+.. In the *Function* box type `count($shiftAssignment)`.
```
Technically "type" should be "enter", per IBM SG (for clarity, supposedly). Another super minor thing. Completely your call. If you do change, be sure to update it everywhere else also.

sterobin commented on docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-shift-required-skills-met-proc.adoc:
```diff
@@ -0,0 +1,32 @@
+. Click *Ok* to create the rule asset.
+. Add a *WHEN* condition by clicking the image:5686.png[] in the *WHEN* field.
+. Select `ShiftAssignment` in the *Add a condition to the rule...* window. Click *+Ok*. 
+. Click on the `ShiftAssignment` condition, and select `employee` from the *Add a restriction on a field* drop down list.
```
Should typically avoid single numbered steps. Just make part of the actual step instead a single sub-step?

sterobin commented on docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-shift-required-skills-met-proc.adoc:
```diff
@@ -0,0 +1,32 @@
+. Click on the `ShiftAssignment` condition, and click *Expression editor*.
+.. In the editor, next to `[not bound]`, select `employee` from the first drop down list, then `skills` from the next drop down list.
+.. Leave the next drop down list as `Choose...`.
+.. In the next drop down box, change `--please choose--` to `excludes`.
```
IMO, I think "change **--please choose--**" is in the same camp as final ellipses in the UI, which are to be deleted (per my previous comment). For that reason, I've just said "change **please choose**" and the like in my own docs, for whatever it's worth. Again, super minor and totally your call.

sterobin reviewed (commented):
Here y' go, @gsheldon . Mostly small, inconsequential things around style and consistency. Nothing that will take long to address, if you choose to.

Commit:
BXMSDOC-2276: Incorporated SME feedback and created planning entity chapter for ShiftAssignment object

Commit:
BXMSDOC-1997: Corrected story title

Commit:
Revert "Revert "BXMSDOC-2246-master Update REBUILT variable for Mar 2 2018 patch publish.""

This reverts commit adb3f0589a24ba74895c1ca4a94e07075f99e9cf.

Commit:
BXMSDOC-2288: Incorporated SME and peer review changes

Commit:
BXMSDOC-2288: Incorporated SME and peer review changes

Conflicts:
	docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-planning-solution-proc.adoc

Commit:
BXMSDOC-2288: Incorporated SME and peer review changes

gsheldon commented:
changes incorporated.

Commit:
Merge pull request #16 from gsheldon/BXMSDOC-2288.backup

Bxmsdoc 2288.backup

Commit:
Revert "Bxmsdoc 2288.backup"

Commit:
Merge pull request #17 from gsheldon/revert-16-BXMSDOC-2288.backup

Revert "Bxmsdoc 2288.backup"

Commit:
Merge pull request #18 from gsheldon/BXMSDOC-2276.backup

Bxmsdoc 2276.backup

Commit:
BXMSDOC-2288: Resolving conflicts
